Upgrade to REL4.0.x

KingRainbow44 opened the pull request:
## Description

Drops support for previous versions of the game, in exchange for upgrading to REL4.0.x

This PR contains multiple bug fixes & questing fixes as well!

## Issues fixed by this PR

<!--- Put the links of issues that may be fixed by this PR here (if any). -->
## Type of changes

<!--- Put an `x` in all the boxes that apply your changes. -->

- [x] Bug fix
- [x] New feature 
- [x] Enhancement
- [x] Documentation

## Checklist:

- [x] My code follows the style guidelines of this project
- [x] My pull request is unique and no other pull requests have been opened for these changes
- [x] I have read the [Contributing note](https://github.com/Grasscutters/Grasscutter/blob/stable/CONTRIBUTING.md) and [Code of conduct](https://github.com/Grasscutters/Grasscutter/blob/development/CODE_OF_CONDUCT.md)
- [x] I am responsible for any copyright issues with my code if it occurs in the future.

KingRainbow44 committed:
Attempt to implement item giving to NPCs (untested)

KingRainbow44 committed:
Add extra giving method

KingRainbow44 committed:
Add command for listing quests

KingRainbow44 committed:
Merge branch 'development' into talking

KingRainbow44 committed:
Add way to toggle certain spam logs

KingRainbow44 committed:
Implement group-based item giving & Add content handler for item giving

KingRainbow44 committed:
Attempt to implement bargaining (untested)

KingRainbow44 committed:
Optimize content handlers

KingRainbow44 committed:
Move more logs to debug constants

KingRainbow44 committed:
Merge remote-tracking branch 'origin/development' into development

KingRainbow44 committed:
Remove deprecated function

KingRainbow44 committed:
Fix errors in chasm

KingRainbow44 committed:
Semi-functional `require` handler implementation

KingRainbow44 committed:
Implement time axis

KingRainbow44 committed:
Implement `ServerLuaCall`

KingRainbow44 committed:
Implement `DebugLog` ability action

KingRainbow44 committed:
Run formatter & bump version

KingRainbow44 committed:
Rename `troubleshoot` to `info` to allow for credits placement

KingRainbow44 committed:
Add new protocol definitions

KingRainbow44 committed:
Add unversioned protocol definitions

KingRainbow44 committed:
Protocol definition fixes

KingRainbow44 committed:
Merge branch 'talking' into dev-4.0

KingRainbow44 committed:
Add `codeSwitch` to the custom config

KingRainbow44 committed:
Fix `BargainRecord`

KingRainbow44 committed:
Add new default avatar abilities

KingRainbow44 committed:
Properly add code switches to the region config

KingRainbow44 committed:
Update packet IDs to 4.0

KingRainbow44 committed:
Update packet files for packet ID changes

hamusuke0323 committed:
feat: ability to register new achievements

hamusuke0323 committed:
fix: liney c3 effect

hamusuke0323 committed:
fix: EntityTeam.java (NPE)

hamusuke0323 committed:
Merge changes from hamusuke (#2)

hamusuke0323 committed:
fix: deobf field name

scooterboo committed:
Make friends by fixing the protos (#2307)

KingRainbow44 committed:
Merge changes from hamusuke (#1)

KingRainbow44 committed:
Run the formatter & Generate protocol definitions

KingRainbow44 committed:
Implement diving

scooterboo committed:
fix forging (again) (#2309)

KingRainbow44 committed:
tamil.

KingRainbow44 committed:
Ensure sensitive information isn't shown without permission

KingRainbow44 committed:
Add alias `troubleshoot` to `info`

KingRainbow44 committed:
Update language files to account for `troubleshoot` -> `info`

scooterboo committed:
Implement Quaternions (#2311)

* Implement Quaternions

* Update BaseRoute.java

---------

Co-authored-by: Magix <27646710+KingRainbow44@users.noreply.github.com>

scooterboo committed:
revert GetScenePointRspOuterClass.java (#2312)

KingRainbow44 committed:
Fix Lua `require`, Fix monster abilities & Fix `ServerLuaCall`

KingRainbow44 committed:
Merge remote-tracking branch 'origin/dev-4.0' into dev-4.0

KiminaNatsuki commented:
he's not responsible for any copyright issues with his code if it occurs in the future

KiminaNatsuki commented:
trust me

KingRainbow44 committed:
Broadcast global value changes

KingRainbow44 committed:
Add `hasnolocals` to resource info

KingRainbow44 committed:
Add debug property to log missing ability actions

KingRainbow44 committed:
Add property for `SetGlobalValue` action

KingRainbow44 committed:
Remove log for 'unimplemented' `SetWeatherAreaState` Lua function

KingRainbow44 committed:
Make `ActionSetGlobalValue` return `true`

KingRainbow44 committed:
Add `PlayerFinishDungeonEvent`

KingRainbow44 committed:
Add `PlayerCompleteQuestEvent`

KingRainbow44 committed:
Add `PlayerForgeItemEvent`

KingRainbow44 committed:
Run formatter on code

KingRainbow44 committed:
Change the result of `call` to a boolean

Returns true if logic execution should continue, false if logic condition should terminate

KingRainbow44 committed:
Add `PlayerCompleteAchievementEvent`

KingRainbow44 committed:
Add `PlayerLevelAvatarEvent`

KingRainbow44 committed:
Add `PlayerLevelItemEvent`

KingRainbow44 committed:
Add `PlayerObtainAvatarEvent`

KingRainbow44 committed:
Add `PlayerObtainItemEvent`

KingRainbow44 committed:
Run formatter on new events

KingRainbow44 committed:
Add `ExecuteCommandEvent`

KingRainbow44 committed:
Run formatter on previous changes

KingRainbow44 committed:
Add `PlayerWishEvent`

KingRainbow44 committed:
Add `PlayerFinishBattlePassMission`

KingRainbow44 committed:
Run formatter on wish event

NotThorny committed:
Bug fixes (#2314)

* Match resources names

* Fix loading with no skill

* Stop overriding common command alias

* Fix discord link in version check popup

* Forgot to add rotation fix

* Remove unnecessary set

hamusuke0323 committed:
just sync it! (#2318)

hamusuke0323 committed:
just sync it! (#2319)

hamusuke0323 committed:
fix: sync team avatar changes (#2320)

KingRainbow44 committed:
Add `PlayerPropertyChangeEvent`

hamusuke0323 committed:
feat: support multiplayer mode in teapot (#2317)

hamusuke0323 committed:
fix: redundant packet 'PacketTakeAchievementRewardReq' in login req (#2321)

hamusuke0323 committed:
fix: player profile changes are not reflected cuz Java is object-oriented (#2322)

* fix: player profile changes are not reflected.
fix: deob FriendBrief.proto

* fix: tentatively sync profile when log out

* fix: allow offline player

* Update src/main/java/emu/grasscutter/game/friends/PlayerProfile.java

Co-authored-by: Magix <27646710+KingRainbow44@users.noreply.github.com>

* Update src/main/java/emu/grasscutter/game/player/Player.java

Co-authored-by: Magix <27646710+KingRainbow44@users.noreply.github.com>

* Update src/main/java/emu/grasscutter/game/player/Player.java

Co-authored-by: Magix <27646710+KingRainbow44@users.noreply.github.com>

* fix: server friend

---------

Co-authored-by: Magix <27646710+KingRainbow44@users.noreply.github.com>

KingRainbow44 committed:
Add `PlayerLevelStatueEvent` & `PlayerUseSkillEvent`

KingRainbow44 committed:
Add `PlayerSwitchAvatarEvent`

KingRainbow44 committed:
Add `PlayerNpcTalkEvent`

hamusuke0323 committed:
feat: add more home mark point (#2323)

KingRainbow44 committed:
Run formatter & 'protogen'

KingRainbow44 committed:
Logging changes (move levels down & add `toString` for some objects)

KingRainbow44 committed:
Compile scripts to replace `require` with script content

KingRainbow44 committed:
De-obfuscate definitions & run formatter

KingRainbow44 committed:
Reflections change & Start HTTP before game server

KingRainbow44 committed:
Merge remote-tracking branch 'origin/dev-4.0' into dev-4.0

# Conflicts:
#	src/main/java/emu/grasscutter/game/player/Player.java
#	src/main/java/emu/grasscutter/game/player/TeamManager.java
#	src/main/java/emu/grasscutter/game/world/Scene.java

KingRainbow44 committed:
Run formatter & update protocol

KingRainbow44 committed:
Disable `LOG_MISSING_ABILITY_HANDLERS`
